### ISSUE TITLE
[Hw-Wallet-Js] Problems with signing

### ISSUE THREAD

**Issue by Senyoret1** [skycoin/hardware-wallet-js#15]
Take this code as an example:

```
const deviceWallet = require('./device-wallet');

deviceWallet.setDeviceType(deviceWallet.DeviceTypeEnum.USB);

let text = "Hello";
//text = "04f42c56cff01496fb71c07c491803ecd3567ef402cd105d5afc085fb109cf42";
//text = "04f42c56cff01496fb71c07c491803ecd3567ef402cd105d5d5afc";

const rejectPromise = function(msg) {
    console.log("Promise rejected", msg);
};
const addrPromise = deviceWallet.devAddressGenPinCode(1, 0);
addrPromise.then(
  (addresses) => {
    console.log("Address Promise resolved", addresses);
    const signPromise = deviceWallet.devSkycoinSignMessagePinCode(0, text);
    signPromise.then(
      (signature) => {
        console.log("Signature promise resolved", signature);
        const checkSignPromise = deviceWallet.devCheckMessageSignature(
          addresses[0],
          text,
          signature
        );
        checkSignPromise.then(
          (check) => {
            console.log(check);
          },
          rejectPromise
        );
      },
      rejectPromise
    );
  },
  rejectPromise
);
```

The code, as it is, works correctly and the console reports that the signature is correct. However, if the line that says `text = "04f42c56cff01496fb71c07c491803ecd3567ef402cd105d5afc085fb109cf42";` is uncommented, the verification fails and the console reports that the signature was made by an address other than the one used. Additionally, if the next line is also uncommented, when running the program for the first time the execution stops and it is necessary to cancel it by pressing ctrl+c, and the second time running the program only part of the instructions is executed. This seems to indicate that there are problems when the messages to sign are too long, something that in practice causes problems, because the hashes that must be signed are 64 characters long.

Additionally, another problem is that the signatures generated by the hardware wallet are different from those used by Skycoin transactions. The hardware wallet creates signatures like `NGV8kPw8FZuYFWYzMa3o...`, which seems to use all the letters of the alphabet, differentiating between upper and lower case, but the Skycoin transactions use hex signatures. It is probably only necessary to convert from one format to another, however, maybe a more profound change is necessary if the signing procedure processes the data in a different way than the Skycoin’s cipher library (I think the cipher library treats the hashes as hex representations of byte arrays, instead of just text strings).

**Comment by mpsido** [comment]
Related to : https://github.com/skycoin/hardware-wallet/issues/35

**Comment by mpsido** [comment]
Your issue raises many problems and misunderstanding @Senyoret1 

## 1. Message format should be a 32 bytes digest 
Normally the messages to sign should be the output of a sha256 digest which is 32 bytes.
But to allow some flexibility here is the cases the firmware is treating
* if the message is not exactly a string of 64 hexadecimal characters, it will hash it for you.
* if it is it will parse the characters and convert them into the 32 bytes we want, then it will sign it for you.

There was a bug in "check signature" indeed that I fixed here: https://github.com/skycoin/hardware-wallet/issues/35. The check signature was "always hashing" which does not work in case you pass something that is "already a hash". 

## 2. A bug in the javascript messages chunking 
As for the "program blocking" it was "my fault", the program split the packets you wanna send in chunks of 64 bytes on the usb wire, but in the message size I failed to count the "header size". I fixed this in commit eee7713.

## 3.  base58 format
You say the base58 format I use for signature it is "different from the Skycoin cipher library", this is not true: You can check here to see the golang conversion between hex and base58 both ways https://github.com/skycoin/skycoin/tree/develop/src/cipher/base58 .
I can create a javascript function to do that for you if you want. But that is another feature. You can create a new issue for that if you want it.
